src/dcqc/filesystems/synapsefs.py

```diff
@@ -0,0 +1,640 @@
+            if self.is_synapse_id(part):
+                children = {entity["id"]: entity for entity in children_list}
+            else:
+                children = {entity["name"]: entity for entity in children_list}
+
+            if part in children:
+                current_entity = children[part]["id"]
+            else:
+                raise ResourceNotFound(path)
+
+        return current_entity
+
+    def _synapse_id_to_entity(self, synapse_id: str, download_file=False) -> Entity:
+        """Retrieve and validate (meta)data for a Synapse entity
+
+        Args:
+            synapse_id (str): A Synapse ID
+            download_file (bool): Whether to download the associated file(s)
+
+        Returns:
+            Entity: The associated Synapse entity
+
+        Raises:
+            fs.errors.ResourceNotFound: If ``synapse_id`` does not exist.
+            fs.errors.ResourceInvalid: If ``synapse_id`` does not correspond
+                 to a supported entity type.
+        """
+        entity = self.synapse.get(synapse_id, downloadFile=download_file)
+        valid_types = list(self.SUPPORTED_TYPES.values())
+        if not any(isinstance(entity, type_) for type_ in valid_types):
+            type_ = type(entity)
+            message = f"{synapse_id} ({type_}) is not supported yet ({valid_types})."
+            raise ResourceInvalid(message)
+        return entity
+
+    def _path_to_entity(self, path: str, download_file=False) -> Entity:
+        """Perform the validation and retrieval steps for a Synapse entity.
+
+        Arguments:
+            path (str): A path.
+            download_file (bool): Whether to download the associated file(s)
+
+        Returns:
+            Entity: A Synapse entity (File, Folder, or Project).
+        """
+        synapse_id = self._path_to_synapse_id(path)
+        with synapse_errors(path):
+            entity = self._synapse_id_to_entity(synapse_id, download_file)
+        return entity
+
+    def _path_to_parent_id(self, path: str) -> str:
+        parent_path, _, basename = path.strip(self.DELIMITER).rpartition(self.DELIMITER)
+        if parent_path == "":
+            if self.root is not None:
+                parent_id = self.root
+            elif self.is_synapse_id(basename):
+                parent_id = self._get_parent_entity(basename)
+            else:
+                message = f"Path ({path}) must start with a Synapse ID when rootless."
+                raise ValueError(message)
+        else:
+            parent_id = self._path_to_synapse_id(parent_path)
+        return parent_id
+
+    def _get_parent_entity(self, entity_id: str) -> str:
+        entity = self.synapse.get(entity_id, downloadFile=False)
+        if isinstance(entity, Project):
+            message = f"Project ({entity.id}) has no parent."
+            raise ValueError(message)
+        parent_id = entity.parentId
+        parent = self.synapse.get(parent_id, downloadFile=False)
+        return parent.id
```

Comment:
You are not returning the parent entity here, but only it's id.  Should this just return `parent_id` or should the function be renamed?

Reply:
Good catch! I changed the return value from the entity to the ID at the last minute, but I forgot to update the method name.